plugins/parsers/avro/parser.go

```diff
@@ -0,0 +1,218 @@
+package avro
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/jeremywohl/flatten/v2"
+	"github.com/linkedin/goavro/v2"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/metric"
+	"github.com/influxdata/telegraf/plugins/parsers"
+)
+
+type Parser struct {
+	MetricName      string            `toml:"metric_name"`
+	SchemaRegistry  string            `toml:"avro_schema_registry"`
+	Schema          string            `toml:"avro_schema"`
+	Measurement     string            `toml:"avro_measurement"`
+	Tags            []string          `toml:"avro_tags"`
+	Fields          []string          `toml:"avro_fields"`
+	Timestamp       string            `toml:"avro_timestamp"`
+	TimestampFormat string            `toml:"avro_timestamp_format"`
+	FieldSeparator  string            `toml:"avro_field_separator"`
+	DefaultTags     map[string]string `toml:"-"`
+
+	Log         telegraf.Logger `toml:"-"`
+	registryObj *SchemaRegistry
+}
+
+func (p *Parser) Parse(buf []byte) ([]telegraf.Metric, error) {
+	schemaID, binaryData, err := p.extractSchemaAndMessage(buf)
+	if err != nil {
+		return nil, err
+	}
+
+	var schema string
+	var codec *goavro.Codec
+
+	if p.SchemaRegistry != "" {
+		schemastruct, err := p.registryObj.getSchemaAndCodec(schemaID)
+		if err != nil {
+			return nil, err
+		}
+		schema = schemastruct.Schema
+		codec = schemastruct.Codec
+	} else {
+		schema = p.Schema
+		codec, err = goavro.NewCodec(schema)
+		if err != nil {
+			return nil, err
+		}
+	}
+	native, _, err := codec.NativeFromBinary(binaryData)
+	if err != nil {
+		return nil, err
+	}
+	// Cast to string-to-interface
+	codecSchema := native.(map[string]interface{})
+	m, err := p.createMetric(codecSchema, schema)
+	if err != nil {
+		return nil, err
+	}
+
+	return []telegraf.Metric{m}, nil
+}
+
+func (p *Parser) extractSchemaAndMessage(buf []byte) (int, []byte, error) {
+	if len(buf) < 5 {
+		err := fmt.Errorf("buf is %d bytes; must be at least 5", len(buf))
+		return 0, nil, err
+	}
+	schemaID := int(binary.BigEndian.Uint32(buf[1:5]))
+	binaryData := buf[5:]
+	return schemaID, binaryData, nil
+}
+
+func (p *Parser) ParseLine(line string) (telegraf.Metric, error) {
+	metrics, err := p.Parse([]byte(line))
+	if err != nil {
+		return nil, err
+	}
+
+	if len(metrics) != 1 {
+		return nil, fmt.Errorf("line contains multiple metrics")
```

Comment:
In case your error does not have any parameter use `errors.New` instead of `fmt.Errorf`
```suggestion
		return nil, errors-New("line contains multiple metrics")
```

Please also check other locations in your code and correct if applicable.